internal/lock.go

```diff
@@ -176,6 +204,72 @@ func (l *Lock) Download(dir string, tags []string, notags []string, perm string)
 	return nil
 }
 
+func downloadFromArtifactory(ctx context.Context, cacheURL string, integrity string, filePath string, mode os.FileMode) error {
+	token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN environment variable is not set")
+	}
+
+	// Extract hash from integrity string
+	h := strings.TrimPrefix(integrity, "sha256-")
+	h = strings.TrimRight(h, "=")
+	padding := len(h) % 4
+	if padding != 0 {
+		h += strings.Repeat("=", 4-padding)
+	}
+
+	hashBytes, err := base64.StdEncoding.DecodeString(h)
+	if err != nil {
+		return fmt.Errorf("failed to decode hash: %v", err)
+	}
+
+	hexHash := hex.EncodeToString(hashBytes)
+	artifactoryURL := fmt.Sprintf("%s/%s", cacheURL, hexHash)
+
+	req, err := http.NewRequestWithContext(ctx, "GET", artifactoryURL, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+token)
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 400 {
+		return fmt.Errorf("failed to download from Artifactory, status code: %d", resp.StatusCode)
+	}
+
+	// If mode is not set, use a default
+	if mode == 0 {
+		mode = 0644
+	}
+
+	// Create a temporary file first
+	tmpFile, err := os.CreateTemp(filepath.Dir(filePath), "download-*")
+	if err != nil {
+		return err
+	}
+	tmpPath := tmpFile.Name()
+	defer os.Remove(tmpPath)
```

Comment:
You're doing a remove... ad on line 270 you're doing a rename... why?

Reply:
You have resolved this without giving me a good explanation... re-opening...

Reply:
I was to trying to make sure the temporary file is deleted if something goes wrong. If everything works fine, the file get renamed and moved to where it needs to be.
Is there a better way to do this?

Reply:
Yes: just remove all this code and re-use the code that exists; you only need to add support for a bearer token:
```
diff --git a/internal/resource.go b/internal/resource.go
index a377e87..aab11b0 100644
--- a/internal/resource.go
+++ b/internal/resource.go
@@ -45,17 +45,20 @@ func NewResourceFromUrl(urls []string, algo string, tags []string, filename stri
 }

 // getUrl downloads the given resource and returns the path to it.
-func getUrl(u string, fileName string, ctx context.Context) (string, error) {
+func getUrl(u string, fileName string, bearer string, ctx context.Context) (string, error) {
 	_, err := url.Parse(u)
 	if err != nil {
 		return "", fmt.Errorf("invalid url '%s': %s", u, err)
 	}
 	log.Debug().Str("URL", u).Msg("Downloading")
-	err = requests.
+	req := requests.
 		URL(u).
 		Header("Accept", "*/*").
-		ToFile(fileName).
-		Fetch(ctx)
+		ToFile(fileName)
+	if bearer != "" {
+		req.Header("Authorization", fmt.Sprintf("Bearer %s", bearer))
+	}
+	err = req.Fetch(ctx)
 	if err != nil {
 		return "", fmt.Errorf("failed to download '%s': %s", u, err)
 	}
@@ -64,12 +67,12 @@ func getUrl(u string, fileName string, ctx context.Context) (string, error) {
 }

 // GetUrlToDir downloads the given resource to the given directory and returns the path to it.
-func GetUrlToDir(u string, targetDir string, ctx context.Context) (string, error) {
+func GetUrlToDir(u string, targetDir string, bearer string, ctx context.Context) (string, error) {
 	// create temporary name in the target directory.
 	h := sha256.New()
 	h.Write([]byte(u))
 	fileName := filepath.Join(targetDir, fmt.Sprintf(".%s", hex.EncodeToString(h.Sum(nil))))
-	return getUrl(u, fileName, ctx)
+	return getUrl(u, fileName, bearer, ctx)
 }

 // GetUrlWithDir downloads the given resource to a temporary file and returns the path to it.
@@ -79,7 +82,7 @@ func GetUrltoTempFile(u string, ctx context.Context) (string, error) {
 		log.Fatal().Err(err)
 	}
 	fileName := file.Name()
-	return getUrl(u, fileName, ctx)
+	return getUrl(u, fileName, "", ctx)
 }

 func (l *Resource) Download(dir string, mode os.FileMode, ctx context.Context) error {
@@ -92,7 +95,7 @@ func (l *Resource) Download(dir string, mode os.FileMode, ctx context.Context) e
 	for _, u := range l.Urls {
 		// Download file in the target directory so that the call to
 		// os.Rename is atomic.
-		lpath, err := GetUrlToDir(u, dir, ctx)
+		lpath, err := GetUrlToDir(u, dir, "", ctx)
 		if err != nil {
 			downloadError = err
 			continue
```